xtractmime/__init__.py

```diff
@@ -1 +1,77 @@
-__version__ = "0.0.0"
+from typing import List, Union, Tuple, Optional
+
+_APACHE_TYPES = [
+    b"text/plain",
+    b"text/plain; charset=ISO-8859-1",
+    b"text/plain; charset=iso-8859-1",
+    b"text/plain; charset=UTF-8",
+]
+
+
+_WHITESPACE_BYTES = {b"\t", b"\r", b"\x0c", b"\n", b" "}
+
+
+# handling resource metadata
+def _should_check_for_apache_bug(supplied_type: Optional[bytes], http_origin: bool) -> bool:
+    return http_origin and supplied_type in _APACHE_TYPES
+
+
+# reading resource header
+def _read_resource_header(body: bytes) -> memoryview:
+    buffer = memoryview(body)
+    if len(body) < 1445:
+        return buffer
+    else:
+        return buffer[:1445]
+
+
+# Matching a MIME type pattern
+def _is_match_mime_pattern(
+    input_bytes: bytes, byte_pattern: bytes, pattern_mask: bytes, lead_whitespace: bool = None
+) -> bool:
+    input_size = len(input_bytes)
+    pattern_size = len(byte_pattern)
+    mask_size = len(pattern_mask)
+
+    if pattern_size != mask_size:
+        raise ValueError("pattern's length should match mask's length")
+
+    if input_size < pattern_size:
+        return False
+
+    input_index, pattern_index = 0, 0
+
+    if lead_whitespace:
+        while (
+            input_index < input_size
+            and input_bytes[input_index : input_index + 1] in _WHITESPACE_BYTES
+        ):
+            input_index += 1
+
+    while pattern_index < pattern_size:
+        masked_byte = bytes([input_bytes[input_index] & pattern_mask[pattern_index]])
+        if masked_byte != byte_pattern[pattern_index : pattern_index + 1]:
+            return False
+        input_index += 1
+        pattern_index += 1
+
+    return True
+
+
+# main function
+def extract_mime(
+    body: bytes,
+    *,
+    content_types: Optional[List[bytes]] = None,
+    http_origin: bool = True,
+    no_sniff: bool = False,
+    extra_types: Optional[List[Tuple[Union[str, bytes]]]] = None,
+) -> str:
+    content_types = content_types if content_types is not None else []
+    extra_types = extra_types if extra_types is not None else []
+    supplied_type = content_types[-1] if content_types else None
```

Comment:
I don’t think you need the first line. And I think the second line can be simplified, to improve readability.

```suggestion
    extra_types = extra_types or tuple()
    supplied_type = content_types[-1] if content_types else None
```